library/core/src/ptr/const_ptr.rs

```diff
@@ -1021,6 +1029,31 @@ impl<T: ?Sized> *const T {
     /// Computes the offset that needs to be applied to the pointer in order to make it aligned to
     /// `align`.
     ///
+    /// Before getting into the precise semantics, it helps to have some context for why this
+    /// operation seems so weird on the surface. This is for two reasons:
+    ///
+    /// * It's for SIMD, specifically [`slice::align_to`]
+    /// * It wants code that uses SIMD operations to work in a `const fn`
+    ///
+    /// In particular, this operation is intended for breaking up a loop into its component
+    /// unaligned and aligned parts, so that the aligned parts can be processed using SIMD.
+    /// For that kind of pattern it's always ok to "fail" to align the pointer, because the
+    /// unaligned path can always handle all the data.
+    ///
+    /// Lots of really basic operations want to use SIMD under the hood, so it would be very
+    /// frustrating if using this pattern made it impossible for an operation to work in
+    /// const contexts. Unfortunately, observing any properties of a pointer's address
+    /// is a very dangerous and problematic operation in const contexts, because it's a huge
+    /// reproducibility issue (which has actual soundness implications with compilation units).
+    ///
+    /// Thankfully, because the precise SIMD pattern we're interested in *already* has a fallback
+    /// mode where the alignment completely fails, the compiler can just make this operation always
+    /// fail to align the pointer when doing const evaluation, and then everything is
+    /// always deterministic and reproducible (and the const evaluator is an interpreter anyway,
+    /// so you weren't actually going to get amazing SIMD speedups).
```

Comment:
More directly: given a slice with a random position in memory, we can't know the head of the slice is aligned, and given a slice, we can have any number of values. Thus we **must** have a head and tail handling routine, because asserting on the size is not good enough (may be off-alignment) and asserting on the alignment is not good enough (may be an uneven number of elements, thus not trivially vectorized without tail-handling), and asserting on both just gives you a panic generator where a useful program should be, because you usually don't really have any way to pre-establish those guarantees in a useful manner based on the type information currently present in the system. And if you actually do, you have much bigger guns than this.

Thus, "you better believe you are going to be writing code to handle this correctly, Miri's semantics be damned."